### ISSUE TITLE
When using the `specgram` function out of the signal pacakge, the plot does not change

### ISSUE THREAD

**Issue by DeepHorizons** [Calysto/octave_kernel#38]
If a spectrogram is created using the `specgram` function on some signal, the plot is generated and displayed. If the signal is then changed and the code re-run, the same plot is shown.

It creates a file in the same folder as the notebook, but deleting this file, clearing the cell output, or restarting the kernel or notebook does not solve the issue.

**Comment by blink1073** [comment]
Can you share a sample of your code?
I tried the following sample and I am able to produce different outputs for different combinations of inputs:

```octave
pkg load signal
x = chirp([0:0.001:2],0,2,500);  # freq. sweep from 0-500 over 2 sec.
Fs=1000;                  # sampled every 0.001 sec so rate is 1 kHz
step=ceil(20*Fs/1000);    # one spectral slice every 20 ms
window=ceil(100*Fs/1000); # 100 ms data window
specgram(x, 2^nextpow2(window), Fs, window, window-step);

**Comment by DeepHorizons** [comment]
If I do
```
x = chirp([0:0.001:2],0,2,500);  # freq. sweep from 0-500 over 2 sec.
Fs=1000;                  # sampled every 0.001 sec so rate is 1 kHz
step=ceil(20*Fs/1000);    # one spectral slice every 20 ms
window=ceil(100*Fs/1000); # 100 ms data window
specgram(x, 2^nextpow2(window), Fs, window, window-step);
```
in one cell and
```
x = chirp([0:0.001:2],0,4,1000);  # freq. sweep from 0-1000 over 4 sec.
specgram(x, 2^nextpow2(window), Fs, window, window-step);
```
in the other, the same graph is shown in both outputs.

![jupyter_octave_but](https://cloud.githubusercontent.com/assets/5551273/13730814/acc011f8-e930-11e5-97be-2929f2c98a07.png)

I noticed that the file it outputs does in fact change. 
An example from a report I'm writing,

![report_example](https://cloud.githubusercontent.com/assets/5551273/13730852/bbbb44ec-e931-11e5-96f5-2e9beffd4f31.png)

The first png output,

![gp_image_01_40](https://cloud.githubusercontent.com/assets/5551273/13730854/c85da2b2-e931-11e5-99ee-c22e689edaa2.png)

And the second png output

![gp_image_01_80](https://cloud.githubusercontent.com/assets/5551273/13730855/ccf50b8a-e931-11e5-877d-c7e8c12f2656.png)

Also, when generating a latex document, it cannot include the plot, instead reporting `Linked image not found`

One question I have is why does it generate a file for specgram and not plots?


I'm using it on JupyterHub on a Linux system.

**Comment by blink1073** [comment]
I'm afraid I don't know how to address your problem.  I can't get the `specgram` to work in my OSX terminal due to display issues, but I can verify that regular SVG outputs work in the notebook, but the ones produced by `specgram` end up as attachments.

**Comment by blink1073** [comment]
Saving as markdown gives a zip file that includes the attached images, if that helps.

**Comment by alexdu** [comment]
Similar issue with plain imshow.

Cell 1:
```
a = randn(500,500);
imshow(a,[]);
```

Cell 2:
```
b = a * 0;
imshow(b,[]);
```

Result:
- both images look like random noise (like the first one)
- the working directory contains only gp_image_01.png, which is black (that's the second image)
- the tool tip in the browser says gnuplot_plot_1a for both images
- opening any of the images in new window (firefox: right click on image, middle click on view image) shows the second image
- after changing the contents of first image, the browser still displays the old image after re-running the cell; refreshing the page helps
- Kernel -> Restart & Run All doesn't help (shows two random noise images in browser, just like when entering the commands manually)

So, there seem to be two separate issues:
- browser caching
- all figures are saved with the same file name

Ubuntu Mate 15.10 (fresh install), octave_kernel 0.15.5, Python 2.7.10, IPython 4.1.2, Firefox 41.0.2.

I can try to fix it, but any hints are welcome. In particular, I found this line:
```
                _filename = fullfile('%s', ['OctaveFig', sprintf('%%03d', _fig)]);
```
but I'm not sure what code loads these files, and also I'm not sure where it's best to store the figure counter, and when to reset it.

**Comment by alexdu** [comment]
I ended up changing the format from SVG to PNG. So far, so good.

**Comment by blink1073** [comment]
Okay, I switched the default to PNG unless on OSX, where PNG is broken, released as `0.15.6`.

**Comment by aisipos** [comment]
I'm getting roughly the same problem on OSX - any 2d type 'image' plot returns the same plot no matter what. I can duplicate on cold start with 2 cells:

`imagesc(rand(5))`

`imagesc(eye(5))`

These will produce the same plot! I noted that in the rendered HTML of the notebook, both showed up as:

`<image height="166.3" preserveAspectRatio="none" width="232.3" x="19.6" xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="gp_image_01.png" y="11.3"></image>`

So PNG is being used, the same filename for each. I also noted that the actual PNG file on disk after running both cells did contain the right image for eye(5), but rendered in the notebook was the same as the first. 

Is there a way I can get these to display as SVG?

**Comment by alexdu** [comment]
My suggestion would be an workaround similar to _fix_gnuplot_svg_size, where you can parse the SVG for gp_image_NN.png and rename the files to get unique names.

From my above links, it doesn't look like gnuplot developers are going to address the issue (though it probably won't hurt to ask them), that's why I suggested the workaround.

**Comment by blink1073** [comment]
I'm not sure exactly how to fix this, but I don't have the bandwidth to investigate at the moment.

**Comment by blink1073** [comment]
@aisipos, I just released a version that should address your problem.  Unfortunately, due to bugs in gnuplot, you have to specify `%plot -f png` when plotting images and `%plot -f svg` when plotting charts.

**Comment by alexdu** [comment]
Looks like plot --size works only with svg backend, but not with png (4c5a162, same OS specs as I've reported above).

BTW, are these options documented anywhere? I didn't know you can change them from the notebook, until seeing your comment (so I was hardcoding them in kernel.py).

**Comment by blink1073** [comment]
`%plot -h` will print out the options.  I'll take a look at png vs svg this weekend.

**Comment by alexdu** [comment]
Thanks.

Just noticed an issue about ```%plot -h```. To reproduce:

1. create a new notebook, with two cells:
- ```imshow(randn(100,100))```
- ```%plot -h```
2. Kernel -> Restart & Run All
3. re-run the first cell with ctrl+enter

Result: octave figure window pops up outside the browser, and the image is not captured in the notebook.

**Comment by blink1073** [comment]
Thanks!  I released metakernel 0.12.6 with the fix.

**Comment by blink1073** [comment]
Closing as fixed.
